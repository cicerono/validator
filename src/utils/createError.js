// @flow
/* eslint-disable lodash-fp/use-fp, lodash-fp/prefer-flat-map */
import {flatten, isArray, keys, map} from "lodash";

export default function createError(
  field: string,
  error: ?string | Array<*>,
  value: any,
  config: Object,
) {
  if (isArray(error)) {
    return flatten(
      map(error, (nestedError, i) =>
        map(keys(nestedError), key =>
          createError(
            `${field}[${i}].${key}`,
            nestedError[key].rule,
            nestedError[key].value,
            nestedError[key].config,
          ),
        ),
      ),
    );
  }

  return {field, rule: error, value, config};
}
