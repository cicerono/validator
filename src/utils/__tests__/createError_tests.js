// @flow
/* eslint-env jest */
import createError from "../createError";

describe("createError()", () => {
  it("should return error object for string", () => {
    const error = createError("field", "numeric", "dd", {});

    expect(error).toEqual({config: {}, field: "field", rule: "numeric", value: "dd"});
  });

  it("should return error object for array", () => {
    const error = createError(
      "array",
      [{key: {field: "key", rule: "numeric", value: "NaN", config: {}}}],
      [{key: "NaN", value: "Value"}],
      {arrayOf: {key: {numeric: {min: 0}}, value: {length: {exact: 1}}}},
    );

    expect(error).toMatchSnapshot();
  });
});
