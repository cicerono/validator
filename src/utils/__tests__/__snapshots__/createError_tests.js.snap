// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createError() should return error object for array 1`] = `
Array [
  Object {
    "config": Object {},
    "field": "array[0].key",
    "rule": "numeric",
    "value": "NaN",
  },
]
`;
